File: BOJ/cpp/15658.cpp
#include <iostream>
#include <algorithm>
#include <string>
#include <vector>
using namespace std;
int n, arr[12], minv=1000000001, maxv=-1000000001;
int op[4]; //+, -, *, /
vector<int> v;

int calc(){
    int result = arr[0];
    for(int i=0; i<v.size(); i++){
        if(v[i] == 0){              // +
            result += arr[i+1];
        }else if(v[i] == 1){        // -
            result -= arr[i+1];
        }else if(v[i] == 2){        // *
            result *= arr[i+1];
        }else if(v[i] == 3){        // /
            result /= arr[i+1];
        }
    }
    return result;
}

void BT(int cnt){
    if(cnt == n-1){
        int result = calc();
        minv = min(minv, result);
        maxv = max(maxv, result);
        return;
    }
    
    for(int i=0; i<4; i++){ //연산자
        if(op[i] != 0){
            op[i]--;
            v.push_back(i); //연산자 번호 넣기 (+:0, -:1, *:2, /:3)
            BT(cnt+1);
            v.pop_back();
            op[i]++;
        }
    }
}

int main(){
    ios::sync_with_stdio(0); cin.tie(0); cout.tie(0);
    cin >> n;
    for(int i=0; i<n; i++) cin >> arr[i];
    for(int i=0; i<4; i++) cin >> op[i];
    
    BT(0);
    cout << maxv << '\n' << minv << '\n';
    
    return 0;
}

File: BOJ/Swift/15658.swift
import Foundation

//input
let n = Int(readLine()!)!
var num = readLine()!.split(separator: " ").map{ Int(String($0))! }
var op = readLine()!.split(separator: " ").map{ Int(String($0))! }
var maxV = -Int.max
var minV = Int.max

func calculator(_ a: Int, _ b: Int, _ op: Int) -> Int {
    switch op {
    case 0: return a + b
    case 1: return a - b
    case 2: return a * b
    case 3: return a / b
    default: return 0
    }
}

func search(_ depth: Int, _ result: Int) { //원하는 개수의 수를 모두 선택한 경우
    if depth == n {
        maxV = max(maxV, result)
        minV = min(minV, result)
        return
    }
    
    for i in 0..<4 {
        if op[i] != 0 {
            op[i] -= 1
            search(depth + 1, calculator(result, num[depth], i))
            op[i] += 1
        }
    }
}

search(1, num[0])
print("\(maxV)\n\(minV)")
